fem/tevaluator.hpp

```diff
@@ -1417,6 +1417,3255 @@ class FieldEvaluator
+   MFEM_ALWAYS_INLINE
+   void AssembleBlock(const qpt_layout_t &qpt_layout,
+                      const qpt_data_t   &qpt_data,
+                      const D_layout_t   &D_layout,
+                      D_data_t           &D_data,
+                      const D11_layout_t &D11_layout,const D11_data_t &D11_data,
+                      const D12_layout_t &D12_layout,const D12_data_t &D12_data,
+                      const D13_layout_t &D13_layout,const D13_data_t &D13_data,
+                      const D21_layout_t &D21_layout,const D21_data_t &D21_data,
+                      const D22_layout_t &D22_layout,const D22_data_t &D22_data,
+                      const D23_layout_t &D23_layout,const D23_data_t &D23_data) const
+   {
+      const int NC = qpt_layout_t::dim_2;
+      const int dof00 = D11_layout_t::dim_1;//D1 == 0 ? DOF2 : DOF1;
+      const int dof01 = D12_layout_t::dim_1;//D1 == 1 ? DOF2 : DOF1;
+      const int dof02 = D13_layout_t::dim_1;//D1 == 2 ? DOF2 : DOF1;
+      const int dof10 = D21_layout_t::dim_2;//D2 == 0 ? DOF2 : DOF1;
+      const int dof11 = D22_layout_t::dim_2;//D2 == 1 ? DOF2 : DOF1;
+      const int dof12 = D23_layout_t::dim_2;//D2 == 2 ? DOF2 : DOF1;
+
+      TTensor4<dof02,NIP*NIP,dof12,NC> A1;
+      TTensor4<dof01,dof02*NIP,dof11,dof12*NC> AA1;
+      OffsetStridedLayout4D<dof00,1,dof01*dof02,dof00,dof10,3*dof00*dof01*dof02*NC,dof11*dof12*NC,3*dof10*dof00*dof01*dof02>
+      M1_layout(D1*(dof00*dof01*dof02*NC)*3*dof10*dof11*dof12*NC
+                +D2*dof10*dof11*dof12*NC);
+      // Using TensorAssemble: <I,NIP,J> --> <DOF,I,DOF,J>
+
+      // qpt_data<NIP1*NIP2,NIP3,NC> --> A1<DOF3,NIP1*NIP2,DOF3,NC>
+      TensorAssemble<false>(
+         D13_layout, D13_data,
+         D23_layout, D23_data,
+         qpt_layout.template split_1<NIP*NIP,NIP>(), qpt_data,
+         A1.layout, A1);
+      // A1<DOF3*NIP1,NIP2,DOF3*NC> --> AA1<DOF2,DOF3*NIP1,DOF2,DOF3*NC>
+      TensorAssemble<false>(
+         D12_layout, D12_data,
+         D22_layout, D22_data,
+         TTensor3<dof02*NIP,NIP,dof12*NC>::layout, A1,
+         AA1.layout, AA1);
+      // AA1<DOF2*DOF3,NIP1,DOF2*DOF3*NC> --> M<DOF1,DOF2*DOF3,DOF1,DOF2*DOF3*NC>
+
+      //TTensor4<DOF1,DOF1*DOF2,DOF1,DOF1*DOF2*NC> M1;
+      TensorAssemble<Add>(
+         D11_layout, D11_data,
+         D21_layout, D21_data,
+         TTensor3<dof01*dof02,NIP,dof11*dof12*NC>::layout, AA1,
+         M1_layout, D_data);
+   }
+
+   // Multi-component assemble of grad-grad element matrices.
+   // qpt_layout is (TNIP x DIM x DIM x NumComp), and
+   // D_layout is (TDOF x TDOF x NumComp).
+   template <typename qpt_layout_t, typename qpt_data_t,
+             typename D_layout_t, typename D_data_t>
+   MFEM_ALWAYS_INLINE
+   void AssembleCurlCurl(const qpt_layout_t &qpt_layout,
+                         const qpt_data_t   &qpt_data,
+                         const D_layout_t   &D_layout,
+                         D_data_t           &D_data) const
+   {
+      //block (0,0)
```

Comment:
Reviewing this part feels a lot like the game of the seven errors. Would it be possible to factorize code, or add comments to understand the general idea... The order in which the blocks are build, what the variables are for instance.